src/bin/miri.rs

```diff
@@ -242,21 +242,49 @@ fn run_compiler(
     callbacks: &mut (dyn rustc_driver::Callbacks + Send),
 ) -> ! {
     if target_crate {
-        // Miri needs a custom sysroot for target crates.
-        // If no `--sysroot` is given, the `MIRI_SYSROOT` env var is consulted to find where
-        // that sysroot lives, and that is passed to rustc.
-        let sysroot_flag = "--sysroot";
-        if !args.iter().any(|e| e == sysroot_flag) {
-            // Using the built-in default here would be plain wrong, so we *require*
-            // the env var to make sure things make sense.
-            let miri_sysroot = env::var("MIRI_SYSROOT").unwrap_or_else(|_| {
-                show_error!(
-                    "Miri was invoked in 'target' mode without `MIRI_SYSROOT` or `--sysroot` being set"
-                    )
-            });
+        if env::var_os("MIRI_STD_AWARE_CARGO").is_some() {
+            let crate_name = parse_crate_name(&args).unwrap_or_else(|| {
+                show_error!("`MIRI_STD_AWARE_CARGO` is set but `--crate-name` is not")
+            })
+            .to_owned();
+            // This list should be kept in sync with the one from
+            // `cargo/core/compiler/standard_lib.rs`.
+            if matches!(
+                crate_name.as_str(),
+                "std"
+                | "core"
+                | "alloc"
+                | "proc_macro"
+                | "panic_abort"
+                | "panic_unwind"
+                | "test"
+                | "compiler_builtins"
+            ) {
+                // We are compiling the standard library.
+                args.push("-Cdebug-assertions=off".into());
+                args.push("-Coverflow-checks=on".into());
+            }
+            // See `cargo_miri::phase_rustc`.
+            if crate_name == "panic_abort" {
+                args.push("-Cpanic=abort".into());
+            }
```

Comment:
Is this really needed? If build-std with regular rustc doesn't need it then we shouldn't need it either. AFAIK build-std will simply only build the panic runtime that is actually needed for this crate and this avoids the issue that comes up when building the panic runtime for a sysroot.